conn.go

```diff
@@ -335,11 +335,11 @@ func (c *Conn) Close() error {
 	return err
 }
 
-// RemoteCertificate exposes the remote certificate
-func (c *Conn) RemoteCertificate() [][]byte {
+// ConnectionState returns basic DTLS details about the connection.
+func (c *Conn) ConnectionState() State {
 	c.lock.RLock()
 	defer c.lock.RUnlock()
-	return c.state.remoteCertificate
+	return c.state
```

Comment:
I think it's better to return `clone()`-ed state to avoid any future data race.

Reply:
Yeah, but that would mean that there is an error that should be checked as well and if I return `(State, error)` this is not the same as crypto/tls.

I've looked at what crypto/tls does, and that also returns a copy of the state, so I think this is good practice to return a copy!

Ideally I think we can do something similar to what crypto/tls is doing, but our state contains other information (e.g. `localRandom`) which is serialised using `.Marshal()`, which could return an error.
So I think eventually we should strip down the connection state to include less stuff so we can copy it like crypto/tls is doing, without possibility of an error. (I'm not sure whether that is possible due to differences between TLS and DTLS, but it would be very nice to have a similar state IMHO.)

So for now, do we return state and error or ignore a possible error?

```go
func (c *Conn) ConnectionState() ConnectionState {
	c.handshakeMutex.Lock()
	defer c.handshakeMutex.Unlock()

	var state ConnectionState
	state.HandshakeComplete = c.handshakeComplete()
	state.ServerName = c.serverName

	if state.HandshakeComplete {
		state.Version = c.vers
		state.NegotiatedProtocol = c.clientProtocol
		state.DidResume = c.didResume
		state.NegotiatedProtocolIsMutual = !c.clientProtocolFallback
		state.CipherSuite = c.cipherSuite
		state.PeerCertificates = c.peerCertificates
		state.VerifiedChains = c.verifiedChains
		state.SignedCertificateTimestamps = c.scts
		state.OCSPResponse = c.ocspResponse
		if !c.didResume && c.vers != VersionTLS13 {
			if c.clientFinishedIsFirst {
				state.TLSUnique = c.clientFinished[:]
			} else {
				state.TLSUnique = c.serverFinished[:]
			}
		}
		if c.config.Renegotiation != RenegotiateNever {
			state.ekm = noExportedKeyingMaterial
		} else {
			state.ekm = c.ekm
		}
	}

	return state
}
```

Reply:
It seems better to rewrite `State.clone()`.
Logically, this clone method should never fail as it is just an equivalent of memory deep copy.
I think adding `handshakeRandom.clone()` resolves it.

Reply:
I'll refactor `State.clone()` quickly.

Reply:
https://github.com/pion/dtls/pull/238

Reply:
Was busy yesterday with other stuff, but I can continue with this today. Thanks for refactoring!